### PR TITLE
add pyamg recipe

### PR DESCRIPTION
**Pull request by jakevdp**
``pyamg`` is part of default conda, but there are no builds for Python 3.X. I [inquired about this on the mailing list](https://groups.google.com/a/continuum.io/forum/#!msg/conda/uM8UNHCznZ4/GZVAK9QYFQAJ) a month ago, but there has been no response.

In light of that, I thought it might make sense to include this in conda forge for the sake of Python 3 users.
Here's my attempt at a pyamg build recipe. Probably needs an additional test script for windows.

### CODE REVIEW

File: recipes/pyamg/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package:
+    name: pyamg
+    version: "3.0.2"
+
+source:
+    git_url: https://github.com/pyamg/pyamg.git
+    git_tag: v3.0.2
+
+build:
+    number: 2
```
Comment (ocefpaf) [inline recipes/pyamg/meta.yaml]: Can you reset this to zero?

File: recipes/pyamg/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    string: np{{CONDA_NPY}}py{{CONDA_PY}}_{{PKG_BUILDNUM}}
```
Comment (ocefpaf) [inline recipes/pyamg/meta.yaml]: We don't encourage the use of build strings. Do you have a strong feeling to add that?

File: recipes/pyamg/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+requirements:
+    build:
+        - python >=2.7,<3|>=3.4,{{PY_VER}}*
```
Comment (ocefpaf) [inline recipes/pyamg/meta.yaml]: Just leave `python` and let `conda-build-all` and `conda-smithy` do the rest for you :wink:

File: recipes/pyamg/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        - numpy {{NPY_VER}}*
```
Comment (ocefpaf) [inline recipes/pyamg/meta.yaml]: Change to `numpy x.x`

File: recipes/pyamg/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package:
+    name: pyamg
+    version: "3.0.2"
+
+source:
+    git_url: https://github.com/pyamg/pyamg.git
+    git_tag: v3.0.2
```
Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: Is there a release we could use here instead of checking out the source code?

File: recipes/pyamg/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    build:
+        - python >=2.7,<3|>=3.4,{{PY_VER}}*
+        - numpy {{NPY_VER}}*
+        - scipy
+        - nose
+        - zlib  # [linux]
+        - gcc 4.8*  # [osx]
+    run:
+        - python {{PY_VER}}*
+        - numpy {{NPY_VER}}*
```
Comment (ocefpaf) [inline recipes/pyamg/meta.yaml]: Same as above.

File: recipes/pyamg/meta.yaml (hunk 4 above)
Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: Yeah, we have special logic for this and it is currently dependent on the `x.x` syntax.

File: recipes/pyamg/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    string: np{{CONDA_NPY}}py{{CONDA_PY}}_{{PKG_BUILDNUM}}
+
+requirements:
+    build:
+        - python >=2.7,<3|>=3.4,{{PY_VER}}*
+        - numpy {{NPY_VER}}*
+        - scipy
+        - nose
+        - zlib  # [linux]
+        - gcc 4.8*  # [osx]
```
Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: Could we not use the system compiler here?

File: recipes/pyamg/run_test.sh [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+if [[ `uname` == Darwin ]] && [ $PY_VER == "2.7" ]; then
+    echo "skipping tests; see https://github.com/pyamg/pyamg/issues/165"
```
Comment (jakirkham) [inline recipes/pyamg/run_test.sh]: Could we add a comment that explains the issue briefly?

File: recipes/pyamg/build.sh [hunk 9]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+# On OSX, we need to ensure we're using conda's gcc/g++
```
Comment (jakirkham) [inline recipes/pyamg/build.sh]: Why is that?

File: recipes/pyamg/meta.yaml (hunk 5 above)
Comment (jakevdp) [inline recipes/pyamg/meta.yaml]: There's a release here: https://pypi.python.org/pypi/pyamg
How would I reference that?

Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: Sure. Take a look at [this]( https://github.com/conda-forge/webcolors-feedstock/blob/master/recipe/meta.yaml#L1-L10 ) as an example.

File: recipes/pyamg/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: Also, could we please add [this]( https://github.com/conda-forge/pytest-benchmark-feedstock/blob/master/recipe/meta.yaml#L14 ) here?

File: recipes/pyamg/bld.bat [hunk 10]
```diff
@@ -0,0 +1,2 @@
+"%PYTHON%" setup.py install
```
Comment (jakirkham) [inline recipes/pyamg/bld.bat]: Could we please update this to match [this]( https://github.com/conda-forge/pytest-benchmark-feedstock/blob/master/recipe/bld.bat#L1 )?

File: recipes/pyamg/build.sh (hunk 9 above)
Comment (jakevdp) [inline recipes/pyamg/build.sh]: Carried over from another recipe – I think it was an OpenMP thing that doesn't actually apply here. I'll remove it.

**Comment by jakevdp** [comment]
@jakirkham – The ``--single-version-externally-managed`` option that you had me add is causing AppVeyor to fail. Should I remove it?

File: recipes/pyamg/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    md5: bd024253ebae6f7bf2d89fe32fc5d64a
+
+build:
+    number: 0
+
+requirements:
+    build:
+        - python
+        - numpy x.x
+        - scipy
```
Comment (ocefpaf) [inline recipes/pyamg/meta.yaml]: Does pymag use setuptools? If not you cannot use `--single-version-externally-managed --record record.txt`

Comment (jakevdp) [inline recipes/pyamg/meta.yaml]: It looks like it uses setuptools optionally: https://github.com/pyamg/pyamg/blob/master/setup.py#L184

Should I add that to build dependencies?

Comment (jakevdp) [inline recipes/pyamg/meta.yaml]: Actually, it only uses setuptools if distutils fails.

Comment (ocefpaf) [inline recipes/pyamg/meta.yaml]: Yes. That is preferred than `distutils`.

File: recipes/pyamg/build.sh [hunk 12]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+$PYTHON setup.py install
```
Comment (ocefpaf) [inline recipes/pyamg/build.sh]: You can remove the `build.sh` file.

File: recipes/pyamg/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    name: pyamg
+    version: {{ version }}
+
+source:
+    fn: pyamg-{{ version }}.tar.gz
+    url: https://pypi.python.org/packages/source/p/pyamg/pyamg-{{ version }}.tar.gz
+    md5: bd024253ebae6f7bf2d89fe32fc5d64a
+
+build:
+    number: 0
```
Comment (ocefpaf) [inline recipes/pyamg/meta.yaml]: You can add here `script: python setup.py install --single-version-externally-managed --record record.txt`

File: recipes/pyamg/meta.yaml (hunk 11 above)
Comment (jakevdp) [inline recipes/pyamg/meta.yaml]: OK – I can remove ``numpy`` as a build requirement, and then it will use setuptools instead. Does that sound reasonable?

Comment (ocefpaf) [inline recipes/pyamg/meta.yaml]: Not sure. I am not familiar with that package to say so. The `numpy x.x` suggests numpy is needed at build time. Maybe it is not. If so you can drop the `x.x` too.

File: recipes/pyamg/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    script: python setup.py install --single-version-externally-managed --record record.txt
+
+requirements:
+    build:
+        - python
+        - setuptools
+        - zlib  # [linux]
+    run:
+        - python
+        - numpy x.x
```
Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: So, does it link against NumPy or not? If so, it should be a build requirement. If not, the `x.x` should be dropped here.

Comment (jakevdp) [inline recipes/pyamg/meta.yaml]: I think it does, but the only way to get it to use setuptools is for numpy to not exist at build time.

I guess this just means it's impossible to use setuptools, and I should remove `` --single-version-externally-managed --record record.txt``, yes?

File: recipes/pyamg/meta.yaml (hunk 13 above)
Comment (jakevdp) [inline recipes/pyamg/meta.yaml]: done

**Comment by jakevdp** [comment]
Is it possible to put the ``nosetests pyamg`` command somewhere in ``meta.yml`` rather than have a separate build script just for this?

**Comment by ccordoba12** [comment]
@jakevdp, sorry if I'm stating the obvious, but it seems `build.sh` and `bld.bat` are missing in your PR.

File: recipes/pyamg/meta.yaml (hunk 13 above)
Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: Yeah, this sadly doesn't work for Windows. So, you may need to bring `bld.bat` back.

File: recipes/pyamg/meta.yaml [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    version: {{ version }}
+
+source:
+    fn: pyamg-{{ version }}.tar.gz
+    url: https://pypi.python.org/packages/source/p/pyamg/pyamg-{{ version }}.tar.gz
+    md5: bd024253ebae6f7bf2d89fe32fc5d64a
+
+build:
+    number: 0
+    script: python setup.py install
```
Comment (ccordoba12) [inline recipes/pyamg/meta.yaml]: I see, you put installation here. Please remove this line and add proper `build.sh` and `bld.bat`.

That's probably why you're getting build errors :-)

File: recipes/pyamg/meta.yaml [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+        - python
+        - numpy x.x
+        - scipy
+        - zlib  # [linux]
+
+test:
+    requires:
+        - nose
+    imports:
+        - pyamg
```
Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: If you want to run a command from `meta.yaml`, just add a `commands` section to `test` and list each command you want to run.

**Comment by jakevdp** [comment]
@ccordoba12 Yes, I removed those at @ocefpaf's suggestion in favor of the ``build: script:`` section of ``meta.yml``

File: recipes/pyamg/meta.yaml (hunk 16 above)
Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: Also, you could use selectors to make sure it doesn't run when it can't.

File: recipes/pyamg/meta.yaml (hunk 15 above)
Comment (jakevdp) [inline recipes/pyamg/meta.yaml]: Originally that's what I had; @ocefpaf asked me to do it this way :smile:

Comment (ocefpaf) [inline recipes/pyamg/meta.yaml]: I asked for the `build.sh` @jakevdp.

@ccordoba12 We don't recommend `build.sh` in conda-forge. And, as soon as `conda-build` support `build/script` for `bld.bat` we'll drop that too.

Comment (jakevdp) [inline recipes/pyamg/meta.yaml]: And the build errors are from the extra ``setup.py`` commands @jakirkham asked me to add, which don't work with numpy's distutils. I've removed them, and now everything seems to be fine.

Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: Yeah, we have been folding in `build.sh` into `meta.yaml`. Though we can't use this in-place of `bld.bat`.

File: recipes/pyamg/meta.yaml (hunk 16 above)
Comment (jakevdp) [inline recipes/pyamg/meta.yaml]: Would this be preferred to having a separate ``run_test.csh``?

File: recipes/pyamg/meta.yaml (hunk 15 above)
Comment (jakevdp) [inline recipes/pyamg/meta.yaml]: So I should un-delete ``bld.bat``?

Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: Correct.

File: recipes/pyamg/meta.yaml (hunk 16 above)
Comment (jakirkham) [inline recipes/pyamg/meta.yaml]: It might be nice. I don't have any strong opinion. @ocefpaf, thoughts?

**Comment by ccordoba12** [comment]
That's odd. I know it's simpler, but we don't do that at Continuum. Besides, we need the `--single-version-externally-managed` option working to avoid egg creation and setuptools messing with `sys.path` (as usual).

@jakevdp, could you revert to those scripts, add `--single-version-externally-managed` and see what happens?

File: recipes/pyamg/meta.yaml (hunk 16 above)
Comment (ocefpaf) [inline recipes/pyamg/meta.yaml]: @jakevdp in that case do whatever is easier for you.  When the test is just 1-line I add it in the `meta.yaml`. When it is more than 1 line I go for a separated script.

**Comment by jakevdp** [comment]
> @jakevdp, could you revert to those scripts, add --single-version-externally-managed and see what happens?

I've tried that several times if you look back through the history of changes here. It causes failures, I think because ``pyamg`` uses numpy/distutils rather than setuptools.

**Comment by ocefpaf** [comment]
> I've tried that several times if you look back through the history of changes here. It causes failures, I think because pyamg uses numpy/distutils rather than setuptools.

Maybe we should patch it.

**Comment by jakevdp** [comment]
Relevant line is here, I believe: https://github.com/pyamg/pyamg/blob/master/setup.py#L184

File: recipes/pyamg/meta.yaml (hunk 16 above)
Comment (jakevdp) [inline recipes/pyamg/meta.yaml]: I'll leave it as-is for now while we figure out the build stuff. No use tweaking more than one thing at a time.

**Comment by jakevdp** [comment]
Things here are quickly getting beyond my level of understanding. If someone with more knowledge of the subtleties of conda builds wants to take over, I'd be happy to let you do that.

**Comment by jakevdp** [comment]
Otherwise I can keep pushing the tweaks you suggest!

**Comment by ocefpaf** [comment]
@jakevdp I am OK merging this without setuptools. We can patch it, and send it upstream, later in the feedstock.

**Comment by ccordoba12** [comment]
@jakevdp, sorry if we are overwhelming you, thanks for keeping up with us :-)

@ocefpaf, I reviewed the build process on Windows and it seems there's no egg created by pyamg after @jakevdp added numpy as a build dep. So there shouldn't be the need for a patch. And since pyamg is using numpy distutils, there's no need for `--single-version-externally-managed` (that's only needed for setuptools).

----

@jakevdp, one last thing: could squash your commits, remove setuptools as a build dep, leave `python setup.py install` in `build: script:`, and remove `bld.bat`?

Thanks a lot :-)

**Comment by jakirkham** [comment]
As long as it passes, I :+1: at this point.

**Comment by ocefpaf** [comment]
No need to squash though. The history here in `staged-recipes` is not that important. The recipe history starts at the feedstock.

Thanks @jakevdp for the recipe.

**Comment by jakirkham** [comment]
Yeah, if they are using any form of `distutils` (without trying `setuptools`), we don't need the extra arguments.

**Comment by jakevdp** [comment]
> remove ``bld.bat``

just to confirm: ``bld.bat`` is not needed after all?

**Comment by jakevdp** [comment]
One final issue is that ``nosetests`` is not run on windows, because it's only in ``run_test.sh``. How should I trigger nosetests for windows?

**Comment by ocefpaf** [comment]
`build.sh` is not needed if you add the `build/script` field with the build command. `bld.bat` **is** still needed b/c conda-build has a bug and ignore the `bld.bat` (for now).

> One final issue is that nosetests is not run on windows, because it's only in run_test.sh. How should I trigger nosetests for windows?

Can you make that test all python and do a `run_test.py`? If not I think that a `run_test.bat` will do.

File: recipes/pyamg/run_test.py [hunk 17]
```diff
@@ -0,0 +1,8 @@
+import nose
+
+if sys.platform == 'darwin' and sys.version.startswith('2.7'):
+    print("There are known errors in pyamg 3.0.2 on Python 2.7 / OSX")
+    print("skipping tests; see https://github.com/pyamg/pyamg/issues/165")
+else:
+    config = nose.config.Config(verbosity=2)
+    nose.runmodule('pyamg', config=config)
```
Comment (jakirkham) [inline recipes/pyamg/run_test.py]: Nice! Definitely a valid alternative to get it to work on all platoforms.

**Comment by jakevdp** [comment]
OK... so the unit test suite is failing occasionally on windows. What's the best thing to do here? I'm already skipping a specific case where OSX has a known failure. Should I do the same for windows?

**Comment by jakirkham** [comment]
Do we know why it is failing? It doesn't seem to have much output. Could we try increasing the verboseness to see if tells us something more about the failure?

**Comment by jakirkham** [comment]
So, I guess some Windows builds do give us [more information]( https://ci.appveyor.com/project/conda-forge/staged-recipes/build/job/sq7826k7nirahy14 ).

**Comment by jakirkham** [comment]
So, I don't know how comfortable you are with Windows. Personally, I rarely use it and don't have a real Windows dev environment. When I am seeing failures on Windows, I end up taking one of the following strategies:

1. Give up and let someone who wants Windows support and knows more try after releasing
    (or possibly try again later myself after the release using 3).
2. Get feedback from someone who knows more and make a few more attempts at it before releasing.
3. Setup my own AppVeyor project for my forked repo and try to test it out there.

Personally, I have no problem with you taking any of these 3 strategies. If it is a new recipe and I am running into problems similar to what you are seeing here, I tend to go for 1. If it is a feedstock recipe that had Windows support before, I try to go for 2 or 3.

**Comment by jakevdp** [comment]
I have very little familiarity or interest in Windows, so I'm going for option 1.

**Comment by jakirkham** [comment]
Sounds good. I'll point you to the changes we need to make, which will be simple.

**Comment by jakevdp** [comment]
I think it's enough to simply skip the test suite – the package builds on windows without a problem.

**Comment by jakevdp** [comment]
Alright, tests pass! I think everything is good to go.

**Comment by ocefpaf** [comment]
Awesome! Thanks for the contribution @jakevdp. Just open an issue regarding the failing tests on Windows when the feedstock is ready.

**Comment by pelson** [comment]
The issue is https://github.com/conda-forge/pyamg-feedstock/issues/1.

Thanks for persevering @jakevdp. If installation is a simple one liner, we tend to prefer using the ```build/script``` section, though that doesn't work on Windows (as raised in https://github.com/conda/conda-build/issues/682).

> pyamg is part of default conda, but there are no builds for Python 3.X. I inquired about this on the mailing list a month ago, but there has been no response.

I'm glad conda-forge could help here. Looks like we have pyamg in the conda-forge channel now (https://anaconda.org/conda-forge/pyamg/files) with py27, py34, py35 x np110, np111 variants. :+1: 

The next step for me is to see if we can get Continuum to take these changes on board and pull the distributions into default. Obviously the pressure is off them, as there is now a very viable solution available in the community. :tada:

**Comment by lukeolson** [comment]
@jakevdp just a heads-up that we're moving to setuptools (in [pyamg/pyamg/tree/setuptools](https://github.com/pyamg/pyamg/tree/setuptools)), hopefully to make this all easier.

**Pull request merged**